File: backend/src/server/streaming_server.py
import os
import threading
import time
from flask import Flask, request, send_file, Response, jsonify, abort, g
from flask_cors import CORS

from monitoring.system_monitor import SystemMonitor
from monitoring.analyzers.connection_analyzer import ConnectionAnalyzer
from security.security_middleware import SecurityMiddleware

class StreamingServerConfig:
    def __init__(self, enableDefense=True, defenseMode='medium', customConfig=None):
        self.enableDefense = enableDefense
        self.defenseMode = defenseMode
        self.customConfig = customConfig or {}

class StreamingServer:
    def __init__(self, config: StreamingServerConfig = StreamingServerConfig()):
        self.config = config
        self.app = Flask(__name__)
        
        # Enable CORS for all routes
        CORS(self.app, resources={
            r"/stream/*": {"origins": "*"},
            r"/download/*": {"origins": "*"},
            r"/security/*": {"origins": "*"},
            r"/health": {"origins": "*"}
        })
        
        self.systemMonitor = SystemMonitor()
        self.connectionAnalyzer = ConnectionAnalyzer()
        self.securityMiddleware = None
        self.isRunning = False

        if self.config.enableDefense:
            self.initialize_defense_system()
        self.setup_routes()

    def initialize_defense_system(self):
        defenseConfig = self.get_defense_config()
        self.securityMiddleware = SecurityMiddleware(self.app, defenseConfig)
        print(f"🛡️ Defense system initialized in {self.config.defenseMode} mode")

    def get_defense_config(self):
        baseConfig = {
            "enableSecurityHeaders": True,
            "enableConnectionThrottling": True,
            "blocklistEnabled": True,
            **self.config.customConfig
        }
        mode = self.config.defenseMode
        if mode == "high":
            return {**baseConfig, **{
                "ackValidationEnabled": True,
                "rateLimitingEnabled": False,
                "sequenceTrackingEnabled": False,
                "adaptiveWindowEnabled": False,
                "anomalyDetectionEnabled": True,
                "quarantineEnabled": True,
                "maxACKsPerSecond": 1000,
                "maxSequenceGap": 10485760,
                "suspiciousPatternThreshold": 0.95,
                "quarantineDuration": 1800000,
                "maxConnectionsPerIP": 50
            }}
        elif mode == "medium":
            return {**baseConfig, **{
                "ackValidationEnabled": True,
                "rateLimitingEnabled": False,
                "sequenceTrackingEnabled": False,
                "adaptiveWindowEnabled": False,
                "anomalyDetectionEnabled": True,
                "quarantineEnabled": True,
                "maxACKsPerSecond": 1000,
                "maxSequenceGap": 10485760,
                "suspiciousPatternThreshold": 0.9,
                "quarantineDuration": 600000,
                "maxConnectionsPerIP": 100
            }}
        elif mode == "low":
            return {**baseConfig, **{
                "ackValidationEnabled": False,
                "rateLimitingEnabled": False,
                "sequenceTrackingEnabled": False,
                "adaptiveWindowEnabled": False,
                "anomalyDetectionEnabled": True,
                "quarantineEnabled": False,
                "maxACKsPerSecond": 10000,
                "maxSequenceGap": 104857600,
                "suspiciousPatternThreshold": 0.99,
                "quarantineDuration": 300000,
                "maxConnectionsPerIP": 200
            }}
        else:  # off
            return {**baseConfig, **{
                "ackValidationEnabled": False,
                "rateLimitingEnabled": False,
                "sequenceTrackingEnabled": False,
                "adaptiveWindowEnabled": False,
                "anomalyDetectionEnabled": False,
                "quarantineEnabled": False,
                "maxACKsPerSecond": 100000,
                "maxSequenceGap": 1048576000,
                "suspiciousPatternThreshold": 1.0,
                "quarantineDuration": 0,
                "maxConnectionsPerIP": 1000
            }}

    def setup_routes(self):
        DATA_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../data'))

        # Debug endpoint for frontend connection testing
        @self.app.route('/debug/cors')
        def debug_cors():
            return jsonify({
                "status": "OK",
                "message": "CORS is working",
                "server": "Optimistic ACK Attack Backend",
                "timestamp": time.time(),
                "origin": request.headers.get('Origin', 'None'),
                "user_agent": request.headers.get('User-Agent', 'None')
            })

        @self.app.route('/download/<filename>')
        def download(filename):
            print(f"🔍 Download request received: {request.path} from {request.remote_addr}")
            print(f"🔍 Headers: {dict(request.headers)}")
            if self.securityMiddleware:
                # download_protection decorator logic
                result = self.securityMiddleware.download_protection(lambda: None)()
                if result: return result

            file_path = os.path.join(DATA_DIR, 'files', filename)
            self.connectionAnalyzer.log_connection(request.remote_addr or 'unknown', 'file_download', filename)

            if not os.path.exists(file_path):
                print(f"File not found: {file_path}")
                abort(404, "File not found")

            # Range support
            range_header = request.headers.get('Range', None)
            stat = os.stat(file_path)
            if range_header:
                start, end = 0, stat.st_size - 1
                match = range_header.replace("bytes=", "").split("-")
                if match[0]:
                    start = int(match[0])
                if len(match) > 1 and match[1]:
                    end = int(match[1])
                chunk_size = end - start + 1
                def generate():
                    with open(file_path, "rb") as f:
                        f.seek(start)
                        remaining = chunk_size
                        while remaining > 0:
                            data = f.read(min(4096, remaining))
                            if not data:
                                break
                            remaining -= len(data)
                            yield data
                rv = Response(generate(), 206, mimetype="application/octet-stream", direct_passthrough=True)
                rv.headers.add('Content-Range', f'bytes {start}-{end}/{stat.st_size}')
                rv.headers.add('Accept-Ranges', 'bytes')
                rv.headers.add('Content-Length', str(chunk_size))
                rv.headers.add('Content-Disposition', f'attachment; filename="{filename}"')
                return rv
            else:
                return send_file(file_path, as_attachment=True)

        @self.app.route('/stream/<streamId>/playlist.m3u8')
        def playlist(streamId):
            if self.securityMiddleware:
                result = self.securityMiddleware.stream_protection(lambda: None)()
                if result: return result
            self.connectionAnalyzer.log_connection(request.remote_addr or 'unknown', 'stream_request', f"playlist-{streamId}")
            playlist = self.generate_hls_playlist(streamId, DATA_DIR)
            return Response(playlist, mimetype="application/vnd.apple.mpegurl")

        @self.app.route('/stream/<streamId>/<segment>')
        def segment(streamId, segment):
            if self.securityMiddleware:
                result = self.securityMiddleware.stream_protection(lambda: None)()
                if result: return result
            self.connectionAnalyzer.log_connection(request.remote_addr or 'unknown', 'stream_request', f"{streamId}/{segment}")
            segment_path = os.path.join(DATA_DIR, 'streams', streamId, segment)
            if os.path.exists(segment_path):
                return send_file(segment_path, mimetype="video/MP2T")
            else:
                abort(404, "Segment not found")

        @self.app.route('/security/metrics')
        def security_metrics():
            if self.securityMiddleware:
                metrics = self.securityMiddleware.get_security_metrics()
                return jsonify(metrics)
            return jsonify({"defenseActive": False})

        @self.app.route('/security/status')
        def security_status():
            return jsonify({
                "defenseActive": self.config.enableDefense,
                "defenseMode": self.config.defenseMode,
                "protectedEndpoints": [
                    '/download/<filename>', '/stream/<streamId>/*'
                ] if self.config.enableDefense else [],
                "lastUpdate": time.strftime("%Y-%m-%dT%H:%M:%S"),
                **({
                    "warning": "Defense system is disabled - server is vulnerable to attacks"
                } if not self.config.enableDefense else {})
            })

    def generate_hls_playlist(self, streamId, DATA_DIR):
        stream_dir = os.path.join(DATA_DIR, 'streams', streamId)
        if not os.path.exists(stream_dir):
            return """#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-ENDLIST"""
        segments = sorted([
            f for f in os.listdir(stream_dir) if f.endswith(".ts")
        ])
        playlist = "#EXTM3U\n#EXT-X-VERSION:3\n#EXT-X-TARGETDURATION:10\n#EXT-X-MEDIA-SEQUENCE:0\n"
        for segment in segments:
            playlist += f"#EXTINF:10.0,\n{segment}\n"
        playlist += "#EXT-X-ENDLIST"
        return playlist

    def start(self, port=8000):
        if not self.isRunning:
            self.isRunning = True
            self.systemMonitor.start()
            if self.config.enableDefense:
                print('🚀 Streaming server started with optimistic ACK attack protection')
                print('🛡️ Defense mechanisms active:')
                print('  ✓ ACK validation (prevents optimistic ACK attacks)')
                print('  ✓ Rate limiting (prevents ACK flooding)')
                print('  ✓ Sequence tracking (detects sequence anomalies)')
                print('  ✓ Window size monitoring (detects abnormal growth)')
                print('  ✓ Anomaly detection (pattern-based detection)')
                print('  ✓ IP quarantine system (automatic blocking)')
                print('  ✓ Connection throttling (prevents resource exhaustion)')
                print(f'  ✓ Defense mode: {self.config.defenseMode.upper()}')
            else:
                print('🚀 Streaming server started WITHOUT PROTECTION')
                print('⚠️  WARNING: Server is vulnerable to attacks!')
                print('   - No ACK validation')
                print('   - No rate limiting')
                print('   - No anomaly detection')
                print('   - No IP blocking')
                print('   Use --defense true to enable protection')
            self.app.run(host="0.0.0.0", port=port, threaded=True)

    def stop(self):
        if self.isRunning:
            self.isRunning = False
            self.systemMonitor.stop()
            if self.securityMiddleware:
                self.securityMiddleware.destroy()
            print('🛑 Streaming server stopped, defense systems deactivated')

    def get_metrics(self):
        base_metrics = {
            "systemMetrics": self.systemMonitor.get_metrics(),
            "connectionMetrics": self.connectionAnalyzer.get_metrics(),
            "isRunning": self.isRunning,
            "defenseEnabled": self.config.enableDefense,
            "defenseMode": self.config.defenseMode
        }
        if self.securityMiddleware:
            return {
                **base_metrics,
                "securityMetrics": self.securityMiddleware.get_security_metrics(),
                "defenseActive": True
            }
        return {
            **base_metrics,
            "securityMetrics": None,
            "defenseActive": False
        }

File: backend/src/server/StreamingServer.ts
import express from 'express';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { SystemMonitor } from '../monitoring/SystemMonitor';
import { ConnectionAnalyzer } from '../monitoring/analyzers/ConnectionAnalyzer';
import { SecurityMiddleware } from '../security/SecurityMiddleware';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface StreamingServerConfig {
  enableDefense: boolean;
  defenseMode: 'high' | 'medium' | 'low' | 'off';
  customConfig?: any;
}

export class StreamingServer {
  private httpServer: express.Application;
  private systemMonitor: SystemMonitor;
  private connectionAnalyzer: ConnectionAnalyzer;
  private securityMiddleware: SecurityMiddleware | null = null;
  private isRunning: boolean = false;
  private config: StreamingServerConfig;

  constructor(config: StreamingServerConfig = { enableDefense: true, defenseMode: 'medium' }) {
    this.config = config;
    this.httpServer = express();
    this.systemMonitor = new SystemMonitor();
    this.connectionAnalyzer = new ConnectionAnalyzer();
    
    if (this.config.enableDefense) {
      this.initializeDefenseSystem();
    }
    
    this.setupRoutes();
  }

  private initializeDefenseSystem(): void {
    const defenseConfig = this.getDefenseConfig();
    this.securityMiddleware = new SecurityMiddleware(defenseConfig);
    this.setupSecurity();
    
    console.log(`🛡️ Defense system initialized in ${this.config.defenseMode} mode`);
  }

  private getDefenseConfig() {
    const baseConfig = {
      enableSecurityHeaders: true,
      enableConnectionThrottling: true,
      blocklistEnabled: true,
      ...this.config.customConfig
    };

    switch (this.config.defenseMode) {
      case 'high':
        return {
          ...baseConfig,
          ackValidationEnabled: true,
          rateLimitingEnabled: false, // Disabled for HTTP
          sequenceTrackingEnabled: false, // Disabled for HTTP
          adaptiveWindowEnabled: false, // Disabled for HTTP
          anomalyDetectionEnabled: true,
          quarantineEnabled: true,
          maxACKsPerSecond: 1000, // High for HTTP
          maxSequenceGap: 10485760, // 10MB - very permissive
          suspiciousPatternThreshold: 0.95, // Very high threshold
          quarantineDuration: 1800000, // 30 minutes
          maxConnectionsPerIP: 50
        };
      
      case 'medium':
        return {
          ...baseConfig,
          ackValidationEnabled: true,
          rateLimitingEnabled: false,
          sequenceTrackingEnabled: false,
          adaptiveWindowEnabled: false,
          anomalyDetectionEnabled: true,
          quarantineEnabled: true,
          maxACKsPerSecond: 1000,
          maxSequenceGap: 10485760,
          suspiciousPatternThreshold: 0.9,
          quarantineDuration: 600000, // 10 minutes
          maxConnectionsPerIP: 100
        };
      
      case 'low':
        return {
          ...baseConfig,
          ackValidationEnabled: false,
          rateLimitingEnabled: false,
          sequenceTrackingEnabled: false,
          adaptiveWindowEnabled: false,
          anomalyDetectionEnabled: true,
          quarantineEnabled: false,
          maxACKsPerSecond: 10000,
          maxSequenceGap: 104857600, // 100MB
          suspiciousPatternThreshold: 0.99,
          quarantineDuration: 300000,
          maxConnectionsPerIP: 200
        };
      
      case 'off':
      default:
        return {
          ...baseConfig,
          ackValidationEnabled: false,
          rateLimitingEnabled: false,
          sequenceTrackingEnabled: false,
          adaptiveWindowEnabled: false,
          anomalyDetectionEnabled: false,
          quarantineEnabled: false,
          maxACKsPerSecond: 100000,
          maxSequenceGap: 1048576000, // 1GB
          suspiciousPatternThreshold: 1.0,
          quarantineDuration: 0,
          maxConnectionsPerIP: 1000
        };
    }
  }

  private setupSecurity(): void {
    if (!this.securityMiddleware) return;
    
    // Apply global security middleware
    this.httpServer.use(this.securityMiddleware.createRequestFilter());
    
    // REMOVED: Aggressive custom rules that were blocking legitimate requests
    // Only add rules for actual attack detection
    
    console.log('🛡️ Security middleware configured - allows legitimate traffic, blocks only attacks');
  }

  private setupRoutes(): void {
    if (this.config.enableDefense) {
      // Protected endpoints with defense middleware
      this.httpServer.get('/download/:filename', 
        (req, res, next) => {
          console.log(`🔍 Download request received: ${req.path} from ${req.ip}`);
          console.log(`🔍 Headers: ${JSON.stringify(req.headers)}`);
          next();
        },
        this.securityMiddleware!.createDownloadProtection(),
        (req, res) => {
          console.log('🔒 Download request passed security validation');
          this.handleFileDownload(req, res);
        });

      this.httpServer.get('/stream/:streamId/playlist.m3u8',
        this.securityMiddleware!.createStreamProtection(),
        (req, res) => {
          this.handleStreamPlaylist(req, res);
        });

      this.httpServer.get('/stream/:streamId/:segment',
        this.securityMiddleware!.createStreamProtection(),
        (req, res) => {
          this.handleStreamSegment(req, res);
        });

      // Security monitoring endpoints
      this.httpServer.get('/security/metrics', (req, res) => {
        const metrics = this.securityMiddleware!.getSecurityMetrics();
        res.json(metrics);
      });

      this.httpServer.get('/security/status', (req, res) => {
        res.json({
          defenseActive: true,
          defenseMode: this.config.defenseMode,
          protectedEndpoints: ['/download/:filename', '/stream/:streamId/*'],
          lastUpdate: new Date().toISOString()
        });
      });
    } else {
      // Unprotected endpoints (original behavior)
      this.httpServer.get('/download/:filename', (req, res) => {
        // console.log('⚠️ Download request - NO PROTECTION ACTIVE');
        this.handleFileDownload(req, res);
      });

      this.httpServer.get('/stream/:streamId/playlist.m3u8', (req, res) => {
        this.handleStreamPlaylist(req, res);
      });

      this.httpServer.get('/stream/:streamId/:segment', (req, res) => {
        this.handleStreamSegment(req, res);
      });

      // Status endpoint showing defense is disabled
      this.httpServer.get('/security/status', (req, res) => {
        res.json({
          defenseActive: false,
          defenseMode: 'off',
          protectedEndpoints: [],
          warning: 'Defense system is disabled - server is vulnerable to attacks',
          lastUpdate: new Date().toISOString()
        });
      });
    }
  }

  private handleFileDownload(req: express.Request, res: express.Response): void {
    const filename = req.params.filename;
    const filePath = path.join(__dirname, '../../data/files', filename);
    
    // Log connection details for monitoring
    this.connectionAnalyzer.logConnection(req.ip || 'unknown', 'file_download', filename);

    if (!fs.existsSync(filePath)) {
      console.error(`File not found: ${filePath}`);
      res.status(404).send('File not found');
      return;
    }

    // Support range requests for resume functionality
    const stat = fs.statSync(filePath);
    const range = req.headers.range;

    if (range) {
      this.handleRangeRequest(res, filePath, stat, range);
    } else {
      res.writeHead(200, {
        'Content-Length': stat.size,
        'Content-Type': 'application/octet-stream',
        'Content-Disposition': `attachment; filename="${filename}"`
      });
      fs.createReadStream(filePath).pipe(res);
    }
  }

  private handleRangeRequest(
    res: express.Response, 
    filePath: string, 
    stat: fs.Stats, 
    range: string
  ): void {
    const parts = range.replace(/bytes=/, "").split("-");
    const start = parseInt(parts[0], 10);
    const end = parts[1] ? parseInt(parts[1], 10) : stat.size - 1;
    const chunksize = (end - start) + 1;

    res.writeHead(206, {
      'Content-Range': `bytes ${start}-${end}/${stat.size}`,
      'Accept-Ranges': 'bytes',
      'Content-Length': chunksize,
      'Content-Type': 'application/octet-stream',
    });

    fs.createReadStream(filePath, { start, end }).pipe(res);
  }

  private handleStreamPlaylist(req: express.Request, res: express.Response): void {
    const streamId = req.params.streamId;
    
    // Log connection for monitoring
    this.connectionAnalyzer.logConnection(req.ip || 'unknown', 'stream_request', `playlist-${streamId}`);
    
    // Generate HLS playlist
    const playlist = this.generateHLSPlaylist(streamId);
    res.setHeader('Content-Type', 'application/vnd.apple.mpegurl');
    res.send(playlist);
  }

  private handleStreamSegment(req: express.Request, res: express.Response): void {
    const { streamId, segment } = req.params;
    const segmentPath = path.join(__dirname, '../../data/streams', streamId, segment);
    
    // Log connection for monitoring
    this.connectionAnalyzer.logConnection(req.ip || 'unknown', 'stream_request', `${streamId}/${segment}`);
    
    if (fs.existsSync(segmentPath)) {
      res.setHeader('Content-Type', 'video/MP2T');
      fs.createReadStream(segmentPath).pipe(res);
    } else {
      res.status(404).send('Segment not found');
    }
  }

  private generateHLSPlaylist(streamId: string): string {
    const streamDir = path.join(__dirname, '../../data/streams', streamId);
    
    // Check if stream directory exists
    if (!fs.existsSync(streamDir)) {
      return `#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-ENDLIST`;
    }

    // Read available segments
    const segments = fs.readdirSync(streamDir)
      .filter(file => file.endsWith('.ts'))
      .sort();

    let playlist = `#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:0
`;

    segments.forEach(segment => {
      playlist += `#EXTINF:10.0,
${segment}
`;
    });

    playlist += '#EXT-X-ENDLIST';
    return playlist;
  }

  public getApp(): express.Application {
    return this.httpServer;
  }

  public start(): void {
    if (!this.isRunning) {
      this.isRunning = true;
      this.systemMonitor.start();
      
      if (this.config.enableDefense) {
        console.log('🚀 Streaming server started with optimistic ACK attack protection');
        console.log('🛡️ Defense mechanisms active:');
        console.log('  ✓ ACK validation (prevents optimistic ACK attacks)');
        console.log('  ✓ Rate limiting (prevents ACK flooding)');
        console.log('  ✓ Sequence tracking (detects sequence anomalies)');
        console.log('  ✓ Window size monitoring (detects abnormal growth)');
        console.log('  ✓ Anomaly detection (pattern-based detection)');
        console.log('  ✓ IP quarantine system (automatic blocking)');
        console.log('  ✓ Connection throttling (prevents resource exhaustion)');
        console.log(`  ✓ Defense mode: ${this.config.defenseMode.toUpperCase()}`);
      } else {
        console.log('🚀 Streaming server started WITHOUT PROTECTION');
        console.log('⚠️  WARNING: Server is vulnerable to attacks!');
        console.log('   - No ACK validation');
        console.log('   - No rate limiting');
        console.log('   - No anomaly detection');
        console.log('   - No IP blocking');
        console.log('   Use --defense true to enable protection');
      }
    }
  }

  public stop(): void {
    if (this.isRunning) {
      this.isRunning = false;
      this.systemMonitor.stop();
      
      if (this.securityMiddleware) {
        this.securityMiddleware.destroy();
      }
      
      console.log('🛑 Streaming server stopped, defense systems deactivated');
    }
  }

  public getMetrics() {
    const baseMetrics = {
      systemMetrics: this.systemMonitor.getMetrics(),
      connectionMetrics: this.connectionAnalyzer.getMetrics(),
      isRunning: this.isRunning,
      defenseEnabled: this.config.enableDefense,
      defenseMode: this.config.defenseMode
    };

    if (this.securityMiddleware) {
      return {
        ...baseMetrics,
        securityMetrics: this.securityMiddleware.getSecurityMetrics(),
        defenseActive: true
      };
    }

    return {
      ...baseMetrics,
      securityMetrics: null,
      defenseActive: false
    };
  }
}
